utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,248 @@
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// struct Foo {
+    ///     str: String,
+    ///     cow: Cow<'static, str>,
+    /// }
+    ///
+    /// fn unsound<'a>(foo: &'a mut Foo) {
+    ///     let a: &str = &foo.str;
+    ///     foo.cow.with_mut(|cow| *cow = Cow::Borrowed(a));
+    /// }
+    /// ```
+    ///
+    /// However, this code will not compile because [`Yokeable::with_mut()`] requires `F: 'static`.
+    /// This enforces that while `F` may mutate `Self<'a>`, it can only mutate it in a way that does
+    /// not insert additional references. For example, `F` may call `to_owned()` on a `Cow` and mutate it,
+    /// but it cannot insert a new _borrowed_ reference because it has nowhere to borrow _from_ --
+    /// `f` does not contain any borrowed references, and while we give it `Self<'a>` (which contains borrowed
+    /// data), that borrowed data is known to be valid
+    ///
+    /// Note that the `for<'b>` is also necessary, otherwise the following code would compile:
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// # use std::mem;
+    ///
+    /// // also safely implements Yokeable<'a>
+    /// struct Bar<'a> {
+    ///     num: u8,
+    ///     cow: Cow<'a, u8>,
+    /// }
+    ///
+    /// fn unsound<'a>(bar: &'a mut Bar<'static>) {
+    ///     bar.with_mut(move |bar| bar.cow = Cow::Borrowed(&bar.num));
+    /// }
+    ///
+    /// # unsafe impl<'a> Yokeable<'a> for Bar<'static> {
+    /// #     type Output = Bar<'a>;
+    /// #     fn transform(&'a self) -> &'a Bar<'a> {
+    /// #         self
+    /// #     }
+    /// #
+    /// #     unsafe fn make(from: Bar<'a>) -> Self {
+    /// #         let ret = mem::transmute_copy(&from);
+    /// #         mem::forget(from);
+    /// #         ret
+    /// #     }
+    /// #
+    /// #     fn with_mut<F>(&'a mut self, f: F)
+    /// #     where
+    /// #         F: 'static + FnOnce(&'a mut Self::Output),
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    ///
+    /// which is unsound because `bar` could be moved later, and we do not want to be able to
+    /// self-insert references to it.
+    ///
+    /// Thus the only types of mutations allowed are ones that move around already-borrowed data, or
+    /// introduce new owned data:
+    ///
+    /// ```rust
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// struct Foo {
+    ///     str: String,
+    ///     cow: Cow<'static, str>,
+    /// }
+    ///
+    /// fn sound<'a>(foo: &'a mut Foo) {
+    ///     foo.cow.with_mut(move |cow| cow.to_mut().push('a'));
+    /// }
+    /// ```
+    fn with_mut<F>(&'a mut self, f: F)
+    where
+        // be VERY CAREFUL changing this signature, it is very nuanced (see above)
+        F: 'static + for<'b> FnOnce(&'b mut Self::Output);
```

Comment:
This universal quantifier is implied, right? (E.g. you'd get the same constraint if you just wrote `FnOnce(&mut Self::Output)`.

 The specific thing that occurs here is that because the argument to F may be of any lifetime, you cannot match it up with `'a`, avoiding the escape described above.

Perhaps the best way to capture this constraint is "F cannot access data that it can prove has particular finite lifetime".

Reply:
It is implied, I'm being explicit about the lifetimes on purpose to avoid rustc picking something for me that I don't like. I'll expand on the comment.

Reply:
+1